### ISSUE TITLE
Stable Diffusion 

### ISSUE THREAD

**Issue by YunchaoYang** [YunchaoYang/Blogs#48]
# Introduction
Stable Diffusion AI is a latent diffusion model for generating AI images.

## Stable Diffusion architecture

Stable Diffusion is a text-to-image open-source model that you can use to create images of different styles and content simply by providing a text prompt. In the context of text-to-image generation, a diffusion model is a generative model that you can use to generate high-quality images from textual descriptions. Diffusion models are a type of generative model that can capture the complex dependencies between the input and output modalities text and images.

The following diagram shows a high-level architecture of a Stable Diffusion model.
![image](https://github.com/YunchaoYang/Blogs/assets/6526592/b46d2927-4136-411e-bf94-063907077486)

It consists of the following key elements:

- **Text encoder** – CLIP is a transformers-based text encoder model that takes input prompt text and converts it into token embeddings that represent each word in the text. CLIP is trained on a dataset of images and their captions, a combination of image encoder and text encoder.
- **U-Net** – A U-Net model takes token embeddings from CLIP along with an array of noisy inputs and produces a denoised output. This happens though a series of iterative steps, where each step processes an input latent tensor and produces a new latent space tensor that better represents the input text.
- **Auto encoder-decoder** – This model creates the final images. It takes the final denoised latent output from the U-Net model and converts it into images that represents the text input.

![image](https://github.com/YunchaoYang/Blogs/assets/6526592/4d637210-1714-4f33-949e-f6e305a30538)

## Types of Stable Diffusion models

<html>
<body>

Model Name | Model Size in GB
-- | --
stabilityai/stable-diffusion-2-1-base | 2.5
stabilityai/stable-diffusion-2-depth | 2.7
stabilityai/stable-diffusion-2-inpainting | 2.5
stabilityai/stable-diffusion-x4-upscaler | 7

</body>
</html>

# Application
### 2. stable diffusion webgui 

running model 1.5

- How to use
```bash
module load python/3.10
wget -q https://raw.githubusercontent.com/AUTOMATIC1111/stable-diffusion-webui/master/webui.sh
# The user environment can be set in webui-user.sh 
# where you can set the python/pytorch executables, etc.
# it will creat venv inside the stable-diffusion-webui folder venv/lib/python3.10
bash webui.sh
# The user interface will pop up in a firefox new web page: 127.0.0.1:7680
```
- evaluate the performance 

## Alternatives of Stable Diffusion

# Solution overview
# Serve models with an NVIDIA Triton Inference Server Python backend

# references
- https://aws.amazon.com/blogs/machine-learning/create-high-quality-images-with-stable-diffusion-models-and-deploy-them-cost-efficiently-with-amazon-sagemaker/
- https://theaisummer.com/diffusion-models/
- https://medium.com/codex/a-quick-look-under-the-hood-of-stable-diffusion-open-source-architecture-2f07fc1e729

**Comment by YunchaoYang** [comment]
# Tips for Stable diffusion image generation
How to get AI generated photos that never capture the true beauty of your real subject? AI generated images typically has certain imagination, which causes randomness.

## How to get prompt

## use ControlNet 

### Model
1. Model 0 
annotation resolution
use openpose get arms 

2. Model 1
First use canny to catch sketch details
then adjust weights for multiple models

And use composable lora and script use 

Sample method Euler A to sample noise overlay
decode steps

4. how to get face with certain expressions
   impainting not masked.

# references
https://www.youtube.com/watch?v=9v4lBexN_Mg

**Comment by YunchaoYang** [comment]
# A number of good prompt examples:
https://github.com/Dalabad/stable-diffusion-prompt-templates/tree/main

**Comment by YunchaoYang** [comment]
AI艺术二维码关键词和参数建议~

大模型：Revanimated: 

Positive prompt： ((best quality)), ((masterpiece)), (detailed), close-up person, long hair, (fantasy art:1.3),  cute cyborg girl, highly detailed face, (render of April:1.1), beautiful artwork illustration, (portrait composition:1.3), (8k resolution:1.2) <lora:Details:1> <lora:IvoryGoldAIv2:0.8>

Negative prompt： (worst quality:1.2), (low quality:1.2), (lowres:1.1), (monochrome:1.1), (greyscale), multiple views, comic, sketch, (((bad anatomy))), (((deformed))), (((disfigured))), watermark, multiple_views, mutation hands, mutation fingers, extra fingers, missing fingers, watermark

大模型：Realistic vision4.0（以下全是positive prompt）negative prompt同上

A beautiful big castle, sunny day, rocket, plane, galaxy, (cloud:1.4), sky, close up shot, by Hayao Miyazaki, trending on artstation, art, 4k, detailed, colorful, bright,  <lora:Details:1>, film grain, studio lighting, HD, highres,
salad, dinner best best quality,  masterpiece, film grain,  <lora:Details:1>
hamburger, dinner best best quality,  masterpiece, film grain,  <lora:Details:1>, film grain, studio lighting
A beautiful landscape in sunshine, elf and wizards, by Hayao Miyazaki, trending on artstation, art, 4k, detailed, colorful, bright,  <lora:Details:1>, film grain, studio lighting
Necronomicon Sketch, page from Necronomicon, crossword puzzle, ((masterpiece),(best quality),(ultra-detailed), trending on artstation,     , film grain, studio lighting, HD, highres,
A beautiful mountain, trending on artstation, art, 4k, detailed, colorful, bright, ,  <lora:Details:1>, film grain, studio lighting
A beautiful man in fancy costume, trending on artstation, art, 4k, detailed, colorful, bright

大模型：XSarchi
Positive prompt: (masterpiece),(high quality), best quality, real,(realistic), super detailed, (full detail),(4k),architecture,Modern style,Blue sky and white clouds <lora:XSArchi_124:0.5>
build,arbor,landscape,water,officialart,extremelydetailedCGunity8kwallpaper

Negative prompt: (normal quality), (low quality), (worst quality), paintings, sketches,fog,signature,soft, blurry,drawing,sketch, poor quality, uply text,type, word, logo, pixelated, low resolution.,saturated,high contrast, oversharpened,dirt 

credit AI 小王子

**Comment by YunchaoYang** [comment]
# 图像风格控制

Stable Diffusion 实现图像风格化的途径主要有以下几种：Artist 艺术家风格、Checkpoint 预训练大模型、LoRA 微调模型、Textual Inversion 文本反转模型。

**Comment by YunchaoYang** [comment]
# LoRA

In diffusion models, especially those like Stable Diffusion, LoRA stands for `Low-Rank Adaptation`. It's a popular technique used for fine-tuning these models efficiently, offering several advantages over other methods. Let's delve deeper into it:

# What it does:

LoRA allows you to adapt a pre-trained Stable Diffusion model to a specific `concept`, `style`, or `character `with significantly smaller file size compared to traditional fine-tuning methods.
It achieves this by introducing only a small number of additional parameters (a few hundred megabytes) instead of modifying the entire model (gigabytes).

## How it works:

Imagine the main model has its weights (W). LoRA adds a smaller set of weight changes (ΔW).
During generation, you can control the influence of these changes using a blending factor (α).
Setting α to 0 uses the original model, and setting it to 1 uses the fully fine-tuned version. Values in between blend their effects.
Why it works:

The key idea is that specific concepts or styles often manifest as subtle adjustments in the model's internal representations. LoRA focuses on capturing these nuanced changes efficiently.
Compared to full fine-tuning, LoRA is:
Faster: Training takes fewer resources and computation time.
Smaller: Resulting models are much lighter and easier to share.
Flexible: You can combine multiple LoRAs for various effects.


While not directly presented in a specific paper, the LoRA technique builds upon several research works:

"Using LoRA for Efficient Stable Diffusion Fine-Tuning" by Hugging Face: This blog post provides a clear explanation and tutorial on using LoRA with Stable Diffusion.
"cloneofsimo/lora" on GitHub: This repository delves into the technical details of LoRA implementation and offers tools for working with it.
Stable Diffusion paper: Understanding the architecture and cross-attention layers of Stable Diffusion (as mentioned in the Hugging Face blog) helps grasp where LoRA intervenes for fine-tuning.
